File: solution/3_Taken/3_Taken/src/Taken.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace taken
{
    /// <summary>
    /// zie leesmij.txt
    /// </summary>
    public class Taken
    {
        private HashSet<Taak> taken = new HashSet<Taak>();

	    public HashSet<Taak> getTaken(){
		    return this.taken;
	    }

	    public HashSet<Taak> geefTakenVanTypeOplossingGestartVoor(TaakType taakType, DateTime oplossingStartDatum) {
		    HashSet<Taak> result = new HashSet<Taak>();
		    foreach (Taak taak in taken) {
			    if (taak.getType() == taakType) {
				    if (taak.getOplossing() != null) {
					    if (DateTime.Compare(taak.getOplossing().getStart().Value, oplossingStartDatum) < 0) {
						    result.Add(taak);
					    }
				    }
			    }
		    }
		    return result;
	    }

	    public List<Taak> geefOpgelosteTakenVanTypeGesorteerdVolgensOplossingsTermijn(TaakType taakType) {
		    List<Taak> result = new List<Taak>();
		    foreach(Taak taak in taken) {
			    if (taak.getType() == taakType) {
				    if (taak.getOplossing() != null && taak.getOplossing().getEind() != null) {
					    result.Add(taak);
				    }
			    }
		    }

            result.Sort(new TaakComparer());
		    return result;
	    }

	    public HashSet<Taak> getGestarteTakenVanType(TaakType taakType) {
		    HashSet<Taak> result = new HashSet<Taak>();
		    foreach(Taak taak in taken) {
			    if (taakType == taak.getType()) {
				    if (taak.getOplossing() != null) {
					    result.Add(taak);
				    }
			    }
		    }
		    return result;
	    }

        private class TaakComparer : IComparer<Taak>
        {
            public int Compare(Taak taak1, Taak taak2)
            {
                return (taak1.getOplossing().getEind().Value.Ticks - taak1.getOplossing().getStart().Value.Ticks).CompareTo(taak2.getOplossing().getEind().Value.Ticks - taak2.getOplossing().getStart().Value.Ticks);
            }
        }
    }
}

File: workspace/3_Taken/src/taken/Taken.java
package taken;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

//zie leesmij.txt
public class Taken {
	
	private Set<Taak> taken = new HashSet<Taak>();

	public Set<Taak> getTaken(){
		return this.taken;
	}

	public Set<Taak> geefTakenVanTypeOplossingGestartVoor(TaakType taakType, Date oplossingStartDatum) {
		Set<Taak> result = new HashSet<Taak>();
		for (Taak taak : taken) {
			if (taak.getType().equals(taakType)) {
				if (taak.getOplossing() != null) {
					if (taak.getOplossing().getStart().before(oplossingStartDatum)) {
						result.add(taak);
					}
				}
			}
		}
		return result;
	}

	public List<Taak> geefOpgelosteTakenVanTypeGesorteerdVolgensOplossingsTermijn(TaakType taakType) {
		List<Taak> result = new ArrayList<Taak>();
		for(Taak taak: taken) {
			if (taak.getType().equals(taakType)) {
				if (taak.getOplossing() != null && taak.getOplossing().getEind() != null) {
					result.add(taak);
				}
			}
		}
		Collections.sort(result, new Comparator<Taak>() {

			@Override
			public int compare(Taak taak1, Taak taak2) {
				return new Long(taak1.getOplossing().getEind().getTime()
						- taak1.getOplossing().getStart().getTime())
						.compareTo(taak2.getOplossing().getEind().getTime()
								- taak2.getOplossing().getStart().getTime());
			}
		});
		return result;
	}

	public Set<Taak> getGestarteTakenVanType(TaakType taakType) {
		Set<Taak> result = new HashSet<Taak>();
		for(Taak taak: taken) {
			if (taakType.equals(taak.getType())) {
				if (taak.getOplossing() != null) {
					result.add(taak);
				}
			}
		}
		return result;
	}

}
